packages/@aws-cdk/aws-iotevents/lib/state.ts

```diff
@@ -28,21 +28,46 @@ export class State {
    */
   public readonly stateName: string;
 
+  private transitionEvents: ITransitionEvent[] = []
+
   constructor(private readonly props: StateProps) {
     this.stateName = props.stateName;
   }
 
   /**
-   * Return the state property JSON
+   * Add a transition event to the state.
+   *
+   * @param transitionEvent the transition event that triggered if condition is evaluated to TRUE
+   */
+  public transitionTo(transitionEvent:ITransitionEvent) {
+    this.transitionEvents.push(transitionEvent);
+  }
+
+  /**
+   * Return the JSON of the states that is transited to.
+   * This function is called recursively and collect the states.
    *
    * @internal
    */
-  public _toStateJson(): CfnDetectorModel.StateProperty {
+  public _getStatesJson(states: CfnDetectorModel.StateProperty[] = []): CfnDetectorModel.StateProperty[] {
     const { stateName, onEnter } = this.props;
-    return {
+
+    if (states.some(s => s.stateName === stateName)) {
+      return states;
```

Comment:
Shouldn't we use a Set here, to check if we haven't see the same state twice?

Also, we're using `stateName` to do the check, but I'm 99% sure we don't do any uniqueness checking on state names, so this is not correct - we should actually compare the State objects.

So, to sum up, this should use a `Set<State>` for checking the duplicates, and return a list of `CfnDetectorModel.StateProperty`. Don't do any accumulation of the result through mutating a variable - instead, just return an empty list if the given State was already visited, and that will handle this case.

Reply:
I'm amaging that codes became clean by this and below advices!